gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/KafkaReportingFormats.java

```diff
@@ -17,87 +17,179 @@
+
+      KafkaAvroEventKeyValueReporter.Builder<?> builder = KafkaAvroEventKeyValueReporter.Factory.forContext(context);
+      if (properties.containsKey(ConfigurationKeys.METRICS_REPORTING_EVENTS_KAFKAPUSHERKEYS)) {
+        List<String> keys = Splitter.on(",").omitEmptyStrings().trimResults()
+            .splitToList(properties.getProperty(ConfigurationKeys.METRICS_REPORTING_EVENTS_KAFKAPUSHERKEYS));
+        builder.withKeys(keys);
+      }
+      if (Boolean.valueOf(properties.getProperty(ConfigurationKeys.METRICS_REPORTING_KAFKA_USE_SCHEMA_REGISTRY,
+          ConfigurationKeys.DEFAULT_METRICS_REPORTING_KAFKA_USE_SCHEMA_REGISTRY))) {
+        builder.withSchemaRegistry(new KafkaAvroSchemaRegistry(properties));
+      }
+      builder.withConfig(getEventsKafkaConfig(properties));
+      String pusherClassName = properties.containsKey(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY_FOR_EVENTS)
+          ? properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY_FOR_EVENTS)
+          : properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY,
+              PusherUtils.DEFAULT_KAFKA_PUSHER_CLASS_NAME);
+      builder.withPusherClassName(pusherClassName);
+
+      return builder.build(brokers, topic);
+
     }
+  },
+  JSON() {
+
+    @Override
+    public void buildMetricsScheduledReporter(String brokers, String topic, Properties properties) throws IOException {
+      KafkaReporter.Builder builder = KafkaReporter.BuilderFactory.newBuilder();
+      builder.build(brokers, topic, properties);
+    }
+
+    @Override
+    public ScheduledReporter buildEventsScheduledReporter(String brokers, String topic, MetricContext context, Properties properties) throws IOException {
+       KafkaEventReporter.Builder builder = KafkaEventReporter.Factory.forContext(context);
+       builder.withConfig(getEventsKafkaConfig(properties));
+       String pusherClassName = properties.containsKey(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY_FOR_EVENTS)
+          ? properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY_FOR_EVENTS)
+          : properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY,
+              PusherUtils.DEFAULT_KAFKA_PUSHER_CLASS_NAME);
+       builder.withPusherClassName(pusherClassName);
+       return builder.build(brokers, topic);
+    }
+  },
+  PLAIN_OBJECT() {
+
+    @Override
+    public void buildMetricsScheduledReporter(String brokers, String topic, Properties properties) throws IOException {
+
+      KafkaKeyValueMetricObjectReporter.Builder<?> builder = KafkaKeyValueMetricObjectReporter.Factory.newBuilder();
+      if (properties.containsKey(ConfigurationKeys.METRICS_REPORTING_EVENTS_KAFKAPUSHERKEYS)) {
+        List<String> keys = Splitter.on(",").omitEmptyStrings().trimResults()
+            .splitToList(properties.getProperty(ConfigurationKeys.METRICS_REPORTING_EVENTS_KAFKAPUSHERKEYS));
+        builder.withKeys(keys);
```

Comment:
This is the tricky point. It's hard to define key in the reporter for a pusher that pushes the key:

- Should we support just one field?
- If we want to support multiple fields, how are they associated, concatenation or interpolate into a pattern?
- Should we support a different object other than any data in the event/metric object?

We shouldn't assume how keys are composed but provide a reasonable default way to compute the key. Here, we made too much assumption. I suggest you move key computation logic in the `report` or `sendEvent` method instead of built-in to the builder.